rapier/src/pipeline.rs

```diff
@@ -1,19 +1,355 @@
 use bevy::app::Events;
 use bevy::ecs::prelude::*;
 use bevy::log::prelude::*;
+use bevy::math::Quat;
 use bevy::math::Vec3;
 use crossbeam::channel::{Receiver, Sender};
 
-use heron_core::{CollisionData, CollisionEvent, Gravity, PhysicsSteps, PhysicsTime};
+use heron_core::{
+    CollisionData, CollisionEvent, CollisionLayers, CollisionShape, Gravity, PhysicsSteps,
+    PhysicsTime,
+};
 
 use crate::convert::{IntoBevy, IntoRapier};
 use crate::rapier::dynamics::{
     CCDSolver, IntegrationParameters, IslandManager, JointSet, RigidBodySet,
 };
 use crate::rapier::geometry::{
-    BroadPhase, ColliderHandle, ColliderSet, ContactEvent, IntersectionEvent, NarrowPhase,
+    BroadPhase, ColliderHandle, ColliderSet, ContactEvent, InteractionGroups, IntersectionEvent,
+    NarrowPhase,
 };
-use crate::rapier::pipeline::{EventHandler, PhysicsPipeline};
+use crate::rapier::parry::query::{Ray, TOIStatus};
+use crate::rapier::pipeline::{EventHandler, PhysicsPipeline, QueryPipeline};
+use crate::shape::ColliderFactory;
+
+use bevy::ecs::system::SystemParam;
+
+pub use physics_world::PhysicsWorld;
+// We have to make a module here so that we can allow missing docs on the structs generated by the
+// derive macro
+#[allow(missing_docs)]
+mod physics_world {
+    #[allow(clippy::wildcard_imports)]
+    // Fine right here because this module is a workaround anyway
+    use super::*;
+
+    /// A Bevy system parameter that can be used to perform queries such as ray casts on the physics
+    /// world
+    ///
+    /// See the [ray_casting](https://github.com/jcornaz/heron/blob/main/examples/ray_casting.rs)
+    /// example for a detailed usage example.
+    #[derive(SystemParam)]
+    pub struct PhysicsWorld<'a> {
+        query_pipeline: ResMut<'a, QueryPipeline>,
+        colliders: ResMut<'a, ColliderSet>,
+    }
+
+    impl<'a> PhysicsWorld<'a> {
+        /// Cast a ray and get the collision shape entity, point, and normal at which it collided,
+        /// if any
+        ///
+        /// - `from`: The point to cast the ray from
+        /// - `to`: The point to cast the ray to
```

Comment:
I think it would be more natural for me to give a direction, rather than a target point.

I also feel that the user will more often have a direction (like "forward") than a target point.

Reply:
:+1: So we can take out the `to` argument, but I see two different ways to do it:

1. We could add a `direction` argument ( a normalized `Vec3` ) and a `distance` argument that would be used to determine which way and how far to cast the ray
2. We could add a `ray` argument ( a `Vec3` ) and use it's direction and distance to determine how far to cast the ray

I'm thinking option 2.